### PR TITLE
chore: add additional information around Cypress configuration

### PR DESCRIPTION
**Pull request by emilyrohrbough**

### CODE REVIEW

File: content/api/commands/url.md [hunk 1]
```diff
@@ -104,11 +104,11 @@ return them the full current URL. We almost never refer to the URL as an `href`.
 
 #### Hardcoded versus using the configuration object
 
-Instead of hardcoding the URL you can use the `baseUrl` of the
-[Cypress configuration](/guides/references/configuration).
+Instead of hard-coding the URL used in the assertion, you can use the `baseUrl` the
+defined in the [Cypress configuration](/guides/references/configuration).
```
Comment (mjhenkes) [inline content/api/commands/url.md:108]: Is one method preferred over the other? Why?

File: content/api/commands/visit.md [hunk 2]
```diff
@@ -42,20 +45,22 @@ cy.visit('./pages/hello.html')
 The URL to visit.
 
 Cypress will prefix the URL with the `baseUrl` configured in your
-[network options](/guides/references/configuration#Global) if you've set one.
+[global configuration](/guides/references/configuration#Global) if you've set one.
```
Comment (mjhenkes) [inline content/api/commands/visit.md:48]: Maybe bette?
```suggestion
[global configuration](/guides/references/configuration#Global) if set.
```

File: content/api/commands/visit.md [hunk 3]
```diff
@@ -224,93 +231,85 @@ cy.visit({
...
 cy.visit('app/index.html')
 ```
 
 #### Visit local file when `baseUrl` is set
 
-If you have `baseUrl` set, but need to visit a local file in a single test or a
-group of tests, disable the `baseUrl` using
-[per-test configuration](/guides/references/configuration#Test-Configuration).
-Imagine our `cypress.json` file:
+If you have `baseUrl` set, but need to visit a local file in a single test or a group of tests, set the `baseUrl` to `null` by using [a test-test configuration override](/guides/references/configuration#Test-time-overrides). Imagine our `cypress.json` file:
```
Comment (mjhenkes) [inline content/api/commands/visit.md:268]:
```suggestion
If you have `baseUrl` set, but need to visit a local file in a single test or a group of tests, set the `baseUrl` to `null` by using [a test-test configuration override](/guides/references/configuration#Test-time-overrides). Given our `cypress.json` file:
```

File: content/api/commands/url.md (hunk 1 above)
Comment (emilyrohrbough) [reply content/api/commands/url.md]: Added a snippet to point to the best practices guide / you-tube video that describe why you should set the baseUrl: [`f3ebeb4` (#4202)](https://github.com/cypress-io/cypress-documentation/pull/4202/commits/f3ebeb460b51d21f6be0b2245b08c68a4a9461e4#diff-6680016ffcad5d4eb739ec1fba3e872c92227b9a4ecea95e2cb018793fa202cdR107)

File: content/api/commands/request.md [hunk 4]
```diff
@@ -39,7 +39,7 @@ cy.visit('http://localhost:8080/app')
 cy.request('users/1.json') //  url is  http://localhost:8080/users/1.json
 ```
 
-2. If you make a `cy.request()` prior to visiting a page, Cypress uses the host
+2. If you make a `cy.request()` prior to visiting a page, Cypress prepends the host
```
[the remark this answers is not in the document]
Comment (emilyrohrbough) [reply content/api/commands/request.md]: Sure! I've read the docs so much at this point, it seems to make sense to me. What do you think is missing?

**Review by mjhenkes — APPROVED** [review]

File: content/api/commands/request.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjhenkes) [reply content/api/commands/request.md]: Looks like you already got this :)

File: content/api/cypress-api/config.md [hunk 5]
```diff
@@ -140,14 +140,14 @@ Cypress.config() // => {defaultCommandTimeout: 10000, viewportHeight: 900, ...}
...
 
-Some configuration values cannot be changed while running a test. Anything
+Some configuration values are readonly and cannot be changed while running a test. Anything
 that's not directly under Cypress's control - like timeouts, `userAgent`, or
-environment variables - will be ignored at run-time.
+environment variables - will be ignored at run-time. Be sure to review the list of
+[test-time configuration options](/guides/references/configuration##Test-time-overrides)
 
-### Test Configuration
+## Test Configuration vs `Cypress.config()`
```
Comment (jennifer-shehane) [inline content/api/cypress-api/config.md:148]: Changing this heading may have broken other links in the docs - need to look for anywhere that might be references `content/api/cypress-api/config#Test-Configuration` and fix all of those links.

File: content/api/cypress-api/config.md [hunk 6]
```diff
@@ -156,8 +156,25 @@ file, using test configuration will only change configuration values during the
 suite or test where they are set. The values will then reset to the previous
 default values after the suite or test is complete.
 
-See the full guide on
-[test configuration](/api/cypress-api/config#Test-Configuration).
+See the full guide on [test-time configuration](/api/cypress-api/config#Test-time-overrides).
```
Comment (jennifer-shehane) [inline content/api/cypress-api/config.md:159]: I don't see this link, where is this?

File: content/api/cypress-api/config.md [hunk 7]
```diff
@@ -156,8 +156,25 @@ file, using test configuration will only change configuration values during the
...
+
+## `Cypress.config()` executes Synchronously
+
+It's important to note that `Cypress.config()` executes synchronously and will not wait for the Cypress commands above it to execute.
+If you need to update your configuration mid-test, be sure to either chain the command the
+[asynchronously Cypress command](/guides/core-concepts/introduction-to-cypress#Commands-Are-Asynchronous) before it or wrap execute it with
+the `cy.then` Cypress command to correctly queue.
```
Comment (jennifer-shehane) [inline content/api/cypress-api/config.md:166]: We don't recommend chaining `.then` directly off of `cy`. If we decide to enforce chaining in a later version this would cause a lot of issues for people we recommended it to.

```suggestion
If you need to update your configuration mid-test, be sure to either chain the command the
[asynchronously using a Cypress command](/guides/core-concepts/introduction-to-cypress#Commands-Are-Asynchronous) before it.
```

File: content/api/cypress-api/config.md [hunk 8]
```diff
@@ -156,8 +156,25 @@ file, using test configuration will only change configuration values during the
...
+
+```javascript
+it('using cy.then', () => {
+  cy.visit('/my-test_page')
+  cy.click('#download-html')
+  cy.then(() => {
+    Cypress.config('baseUrl', 'null')
+  })
```
Comment (jennifer-shehane) [inline content/api/cypress-api/config.md:174]:
```suggestion
  cy.click('#download-html')
    .then(() => {
      Cypress.config('baseUrl', 'null')
    })
```

File: content/guides/references/configuration.md [hunk 9]
```diff
@@ -330,35 +330,10 @@ You can
 
 </Alert>
 
-### `Cypress.config()`
+## Test-time overrides
```
Comment (jennifer-shehane) [inline content/guides/references/configuration.md:333]: I'm not sure we want to update the wording of this to 'Test-time overrides'. This change would require consulting with product/DX as this introduces brand new terminology to the product.

File: content/guides/core-concepts/writing-and-organizing-tests.md [hunk 10]
```diff
@@ -502,58 +502,13 @@ it.skip('returns "fizz" when number is multiple of 3', () => {
 
 ### Test Configuration
 
-To apply a specific Cypress [configuration](/guides/references/configuration)
-value to a suite or test, pass a configuration object to the test or suite
-function as the second argument.
+It is possible to apply [test-time configuration](/guides/references/configuration#Test-time-overrides) values during test execution.
```
Comment (jennifer-shehane) [inline content/guides/core-concepts/writing-and-organizing-tests.md:505]: I would prefer to keep this section in this document. The configuration document is pretty overloaded as it is. Any reasoning for moving this section?

File: content/guides/references/best-practices.md [hunk 11]
```diff
@@ -913,13 +913,27 @@ We have
...
 </Alert>
 
-Adding a [baseUrl](/guides/references/configuration#Global) in your
-configuration allows you to omit passing the `baseUrl` to commands like
-[cy.visit()](/api/commands/visit) and [cy.request()](/api/commands/request).
-Cypress assumes this is the url you want to use.
+By adding a [baseUrl](/guides/references/configuration#Global) in your configuration
+Cypress will attempt to prefix it any url provided to commands like [cy.visit()](/api/commands/visit)
+and [cy.request()](/api/commands/request) that are not fully qualified domain name (FQDN) urls.
+
```
Comment (jennifer-shehane) [inline content/guides/references/best-practices.md:919]:
```suggestion
By adding a [baseUrl](/guides/references/configuration#Global) in your configuration
Cypress will attempt to prefix the `baseUrl` to any URL provided to commands like [cy.visit()](/api/commands/visit)
and [cy.request()](/api/commands/request) that are not fully qualified domain name (FQDN) urls.

```

File: content/guides/references/best-practices.md [hunk 12]
```diff
@@ -913,13 +913,27 @@ We have
...
+This allows you to omit hard-coding fully qualified domain name (FQDN) urls
```
Comment (jennifer-shehane) [inline content/guides/references/best-practices.md:920]:
```suggestion
This allows you to omit hard-coding fully qualified domain name (FQDN) URLs
```

File: content/guides/references/best-practices.md [hunk 13]
```diff
@@ -913,13 +913,27 @@ We have
...
+
+can be shortened to
+
+```javascript
+ cy.visit('index.html')
+```
+
+Not only does this create tests that can easily flex between domains, i.e. running a dev server on
+`http:localhost:8080` vs a deployed production server domain, but adding a baseUrl can also save some time
+during the initial startup of your Cypress tests.
```
Comment (jennifer-shehane) [inline content/guides/references/best-practices.md:935]: I'm not sure the meaning of 'flex' is clear here. 

```suggestion
Not only does this create tests that can easily switch between domains, i.e. running a dev server on
`http:localhost:8080` vs a deployed production server domain, but adding a `baseUrl` can also save some time
during the initial startup of your Cypress tests.
```

File: content/guides/references/best-practices.md [hunk 14]
```diff
@@ -957,3 +971,7 @@ We also display an error if your server is not running at the specified
 `baseUrl` during `cypress run` after several retries.
 
 <DocsImage src="/img/guides/cypress-verifies-server-is-running-during-cypress-run.png" alt="The terminal warns and retries when the url at your baseUrl is not running" ></DocsImage>
+
+### Usage of baseUrl in depth
+
+This [short video](https://www.youtube.com/watch?v=f5UaXuAc52c) explains in depth how to use baseUrl correctly.
```
Comment (jennifer-shehane) [inline content/guides/references/best-practices.md:977]:
```suggestion
This [short video](https://www.youtube.com/watch?v=f5UaXuAc52c) explains in depth how to use `baseUrl` correctly.
```

File: content/guides/references/best-practices.md [hunk 15]
```diff
@@ -957,3 +971,7 @@ We also display an error if your server is not running at the specified
 `baseUrl` during `cypress run` after several retries.
 
 <DocsImage src="/img/guides/cypress-verifies-server-is-running-during-cypress-run.png" alt="The terminal warns and retries when the url at your baseUrl is not running" ></DocsImage>
+
+### Usage of baseUrl in depth
```
Comment (jennifer-shehane) [inline content/guides/references/best-practices.md:975]:
```suggestion
### Usage of `baseUrl` in depth
```

File: content/guides/references/configuration.md [hunk 16]
```diff
@@ -330,35 +330,10 @@ You can
...
-### Test Configuration
-
-To apply specific Cypress [configuration](/guides/references/configuration)
-values to a suite or test, pass a configuration object to the test or suite
-function as the second argument.
-
-The configuration values passed in will only take effect during the suite or
-test where they are set. The values will then reset to the previous default
-values after the suite or test is complete.
+We provide two options for override the configuration while your test are running, `Cypress.config()` and
```
Comment (jennifer-shehane) [inline content/guides/references/configuration.md:335]:
```suggestion
We provide two options to override the configuration while your test are running, `Cypress.config()` and
```

File: content/guides/references/configuration.md [hunk 17]
```diff
@@ -383,57 +359,90 @@ configuration:
...
+#### `Cypress.config()`
 
-You can configure the number of times to retries a suite of tests if they fail
-during `cypress run` and `cypress open` separately.
+You can also override configuration values within your test using
+[`Cypress.config()`](/api/cypress-api/config).
+
+
+This changes the configuration _for the remaining execution of the current spec file_.
+The values will reset to the previous default values after spec has complete.
```
Comment (jennifer-shehane) [inline content/guides/references/configuration.md:369]:
```suggestion
The values will reset to the previous default values after the spec has complete.
```

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
@emilyrohrbough Thanks for the updates. It'd be great to understand your reasoning more behind the changes. When you reopen this PR, can you put a thorough description/comments on why you felt these changes were necessary. - Like, 'I was really confused at this point, or it made more sense for this portion to be here because'

File: content/api/commands/visit.md [hunk 18]
```diff
@@ -42,20 +45,22 @@ cy.visit('./pages/hello.html')
...
+If the the baseUrl has not been set, you will need to specify a fully qualified url or Cypress will attempt to
+act as your web server. See the [prefixes notes](#prefixes) for more details.
+
+**Note:** visiting a new domain requires the Test Runner window to reload. You cannot visit different domains in a single test.
 
 **<Icon name="angle-right"></Icon> options** **_(Object)_**
 
 Pass in an options object to control the behavior of `cy.visit()`.
 
+By default, the `cy.visit()` commands' will use the `pageLoadTimeout` and `baseUrl` set globally in your {% url 'configuration' configuration %}.
```
[the remark this answers is not in the document]
Comment (emilyrohrbough) [reply content/api/commands/visit.md:59]: These aren't documented anywhere that I am aware of. I had made these documentation updates before the new vue changes had been merged in.

File: content/api/cypress-api/config.md (hunk 5 above)
Comment (emilyrohrbough) [reply content/api/cypress-api/config.md:148]: There are no links to `content/api/cypress-api/config#Test-Configuration`

File: content/guides/references/configuration.md (hunk 9 above)
Comment (emilyrohrbough) [reply content/guides/references/configuration.md:333]: I'll change

File: content/api/commands/visit.md [hunk 19]
```diff
@@ -224,93 +231,85 @@ cy.visit({
...
   cy.visit('index.html')
   cy.contains('local file')
 })
 ```
 
-**Tip:** because visiting every new domain requires the Test Runner window
-reload, we recommend putting the above tests in separate spec files.
-
-### Prefixes
+**Tip:** because visiting every new domain requires the Test Runner window reload, we recommend putting the above tests in separate spec files.
```
[the remark this answers is not in the document]
Comment (emilyrohrbough) [reply content/api/commands/visit.md:289]: removed.

File: content/guides/core-concepts/writing-and-organizing-tests.md (hunk 10 above)
Comment (emilyrohrbough) [reply content/guides/core-concepts/writing-and-organizing-tests.md:505]: I had removed this section since it repeated the explanation & examples snippet that are already provided in `/guides/references/configuration`. This is one less place to check for configuration documentation changes. I can re-add.
